tasks/migrate-removals.ts

```diff
@@ -245,36 +283,84 @@ export const GET_MIGRATE_REMOVALS_TASK = () =>
           dryRun === true
             ? removalContract.callStatic.mintBatch
             : removalContract.mintBatch;
-        let pendingTx: Awaited<ReturnType<typeof migrationFunction>>;
 
+        // Enable this code to simulate a timeout on third project to be minted on localhost
+        // if (projectIndex === 2 && network === 'localhost') {
+        //   logger.info(`🚧 Intentionally timing out on third project`);
+        //   migrationFunction = async () => {
+        //     console.log('Calling the timeout fake mintBatch function...');
+        //     await new Promise((resolve) =>
+        //       // eslint-disable-next-line no-promise-executor-return -- script
+        //       setTimeout(resolve, TIMEOUT_DURATION * 2)
+        //     ); // will time out
+        //   };
+        // }
+
+        let pendingTx: ContractTransaction;
+        let tokenIds;
+        let txReceipt;
         try {
-          const gasPrice = await signer.getGasPrice(); // TODO on a live network we probably don't need to do this??
-          pendingTx = await removalContract.mintBatch(
-            signerAddress, // mint to the consignor
-            amounts,
-            removals,
-            project.projectId,
-            project.scheduleStartTime,
-            project.holdbackPercentage,
-            { gasPrice }
-          );
-          const txResult = await pendingTx.wait(1); // TODO specify more than one confirmation?
-          const txReceipt =
-            await removalContract.provider.getTransactionReceipt(
-              txResult.transactionHash
+          let maybePendingTx;
+          if (network === `localhost` && dryRun === false) {
+            // localhost non-dry-run requires manually setting gas price
+            const gasPrice = await signer.getGasPrice();
+            maybePendingTx = await callWithTimeout(
+              migrationFunction(
+                signerAddress, // mint to the consignor
+                amounts,
+                removals,
+                project.projectId,
+                project.scheduleStartTime,
+                project.holdbackPercentage,
+                { gasPrice }
+              ),
+              TIMEOUT_DURATION
             );
-          const tokenIds = parseTransactionLogs({
-            contractInstance: removalContract,
-            txReceipt,
-            eventNames: ['TransferBatch'],
-          }).flatMap((log) =>
-            log.args.ids.map((id: BigNumber) => id.toString())
-          );
-          if (txReceipt.status !== 1) {
-            logger.error(
-              `❌ Transaction ${pendingTx.hash} failed with failure status ${txReceipt.status} - exiting early`
+          } else {
+            // all other cases
+            maybePendingTx = await callWithTimeout(
+              migrationFunction(
```

Comment:
Are you _**sure**_ this works as expected? When you use a promise, **_even if you don't await it_**, it starts to execute

Reply:
to avoid that, pass it as a callback instead (`()=> fn()`) or use the provider timeout setting I linked before

Reply:
fixed this using the callback approach.  this was working because it was simply meant to make sure that a promise doesn't hang forever, but I see how passing a promise that has already begun execution means that it may end up running longer than the timeout before things actually time out, which could be unexpected behavior in other applications of a function like this.